// @react95/core/cjs/Video/Video.css.cjs
require('./Video.css.ts.vanilla.css');
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
/* empty css                             */const createRuntimeFn = require("@vanilla-extract/recipes/createRuntimeFn");
var videoTag = createRuntimeFn.createRuntimeFn({ defaultClassName: "r95_13gnpv00", variantClassNames: { visible: { true: "r95_13gnpv01", false: "r95_13gnpv02" } }, defaultVariants: {}, compoundVariants: [] });
var controls = "r95_13gnpv03";
var countDownContainer = "r95_13gnpv04";
var videoFont = "r95_13gnpv05";
var currentTime = "r95_13gnpv06";
var elapsedTime = "r95_13gnpv07";
var divider = "r95_13gnpv08";
var controlBtn = "r95_13gnpv09";
var range = "r95_13gnpv0a";
exports.controlBtn = controlBtn;
exports.controls = controls;
exports.countDownContainer = countDownContainer;
exports.currentTime = currentTime;
exports.divider = divider;
exports.elapsedTime = elapsedTime;
exports.range = range;
exports.videoFont = videoFont;
exports.videoTag = videoTag;
