// @react95/core/cjs/TaskBar/Clock.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const React = require("react");
const Frame = require("../Frame/Frame.cjs");
const Tooltip = require("../Tooltip/Tooltip.cjs");
const TaskBar_css = require("./TaskBar.css.cjs");
const Clock = () => {
  const [timer, setTimer] = React.useState("");
  React.useEffect(() => {
    function checkTime(i) {
      return i < 10 ? `0${i}` : i;
    }
    const interval = setInterval(() => {
      const today = /* @__PURE__ */ new Date();
      const h = today.getHours();
      const m = today.getMinutes();
      setTimer(`${checkTime(h)}:${checkTime(m)}`);
    });
    return () => clearInterval(interval);
  }, []);
  return /* @__PURE__ */ React.createElement(
    Frame.Frame,
    {
      boxShadow: "$in",
      px: "$6",
      py: "$2",
      display: "flex",
      justifyContent: "center",
      alignItems: "center"
    },
    /* @__PURE__ */ React.createElement(Tooltip.Tooltip, { className: TaskBar_css.tooltip }, timer)
  );
};
exports.Clock = Clock;
