// @react95/core/esm/themes.mjs
import * as azureOrange_css from "./themes/azureOrange.css.mjs";
import * as bee_css from "./themes/bee.css.mjs";
import * as blackAndWhite_css from "./themes/blackAndWhite.css.mjs";
import * as brick_css from "./themes/brick.css.mjs";
import * as candy_css from "./themes/candy.css.mjs";
import * as coldGray_css from "./themes/coldGray.css.mjs";
import * as counterStrike_css from "./themes/counterStrike.css.mjs";
import * as eggplant_css from "./themes/eggplant.css.mjs";
import * as highContrast_css from "./themes/highContrast.css.mjs";
import * as lilac_css from "./themes/lilac.css.mjs";
import * as lilacRoseDark_css from "./themes/lilacRoseDark.css.mjs";
import * as maple_css from "./themes/maple.css.mjs";
import * as marine_css from "./themes/marine.css.mjs";
import * as matrix_css from "./themes/matrix.css.mjs";
import * as millenium_css from "./themes/millenium.css.mjs";
import * as modernDark_css from "./themes/modernDark.css.mjs";
import * as molecule_css from "./themes/molecule.css.mjs";
import * as ninjaTurtles_css from "./themes/ninjaTurtles.css.mjs";
import * as olive_css from "./themes/olive.css.mjs";
import * as pamelaAnderson_css from "./themes/pamelaAnderson.css.mjs";
import * as plum_css from "./themes/plum.css.mjs";
import * as rainyDay_css from "./themes/rainyDay.css.mjs";
import * as rose_css from "./themes/rose.css.mjs";
import * as slate_css from "./themes/slate.css.mjs";
import * as spruce_css from "./themes/spruce.css.mjs";
import * as storm_css from "./themes/storm.css.mjs";
import * as theSixtiesUSA_css from "./themes/theSixtiesUSA.css.mjs";
import * as tokyoDark_css from "./themes/tokyoDark.css.mjs";
import * as tooSexy_css from "./themes/tooSexy.css.mjs";
import * as travel_css from "./themes/travel.css.mjs";
import * as vaporTeal_css from "./themes/vaporTeal.css.mjs";
import * as vermillion_css from "./themes/vermillion.css.mjs";
import * as violetDark_css from "./themes/violetDark.css.mjs";
import * as water_css from "./themes/water.css.mjs";
import * as win95_css from "./themes/win95.css.mjs";
export {
  azureOrange_css as azureOrange,
  bee_css as bee,
  blackAndWhite_css as blackAndWhite,
  brick_css as brick,
  candy_css as candy,
  coldGray_css as coldGray,
  counterStrike_css as counterStrike,
  eggplant_css as eggplant,
  highContrast_css as highContrast,
  lilac_css as lilac,
  lilacRoseDark_css as lilacRoseDark,
  maple_css as maple,
  marine_css as marine,
  matrix_css as matrix,
  millenium_css as millenium,
  modernDark_css as modernDark,
  molecule_css as molecule,
  ninjaTurtles_css as ninjaTurtles,
  olive_css as olive,
  pamelaAnderson_css as pamelaAnderson,
  plum_css as plum,
  rainyDay_css as rainyDay,
  rose_css as rose,
  slate_css as slate,
  spruce_css as spruce,
  storm_css as storm,
  theSixtiesUSA_css as theSixtiesUSA,
  tokyoDark_css as tokyoDark,
  tooSexy_css as tooSexy,
  travel_css as travel,
  vaporTeal_css as vaporTeal,
  vermillion_css as vermillion,
  violetDark_css as violetDark,
  water_css as water,
  win95_css as win95
};
